### ISSUE TITLE
Test cases fail on Debian/Ubuntu

### ISSUE THREAD

**Issue by bdrung** [libgit2/libgit2#4186]
I tried to update the Debian package of libgit2, but some test cases fail on Debian testing:

```
  1) Failure:
core::features::0 [/<<PKGBUILDDIR>>/tests/core/features.c:20]
  Expression is not true: (caps & GIT_FEATURE_HTTPS) != 0

  2) Failure:
repo::discover::discovering_repository_with_alternative_gitdir_succeeds [/<<PKGBUILDDIR>>/tests/repo/discover.c:36]
  Function call failed: (git_repository_discover(&found_path, start_path, 0, ceiling_dirs))
  error -3 - could not find repository from 'temprepo/discover.git/alternate_sub_repo/sub'

  3) Failure:
repo::discover::ceiling_should_not_affect_gitdir_redirection [/<<PKGBUILDDIR>>/tests/repo/discover.c:36]
  Function call failed: (git_repository_discover(&found_path, start_path, 0, ceiling_dirs))
  error -3 - could not find repository from 'temprepo/discover.git/alternate_sub_repo/sub'

  4) Failure:
repo::discover::discovery_starting_at_file_succeeds [/<<PKGBUILDDIR>>/tests/repo/discover.c:36]
  Function call failed: (git_repository_discover(&found_path, start_path, 0, ceiling_dirs))
  error -3 - could not find repository from 'temprepo/discover.git/sub_repo/file'

  5) Failure:
repo::open::force_bare [/<<PKGBUILDDIR>>/tests/repo/open.c:401]
  Function call failed: (git_repository_open_ext( &barerepo, "alternate/subdir/sub2", GIT_REPOSITORY_OPEN_BARE, ((void *)0)))
  error -3 - could not find repository from 'alternate/subdir/sub2'
```

Failure 2 to 5 can be only seen when running the tests inside sbuild (which uses schroot and overlay directories in my case).

### Reproduction steps

1. Download the packaging from Debian and extract it into the directory `debian` in your libgit2 directory:
http://http.debian.net/debian/pool/main/libg/libgit2/libgit2_0.24.5-1.debian.tar.xz
2. run `dpkg-buildpackage -b` (or better use sbuild)

### Version of libgit2 (release number or SHA1)

Version 0.25.1 and current head (caf7a7a6b14b3b2f200f88f2f8d7a18ca10fd3ac) are affected.

### Operating system(s) tested

Debian testing and Ubuntu 16.10 (yakkety)

**Comment by pks-t** [comment]
I was able to reproduce the HTTPS test failure, which should be fixed by #4216. But I was not able to reproduce the error with repository discovery when executing `dpkg-buildpackage -b`. I didn't try sbuild cause I'm unfamiliar with the Debian build infrastructure and sbuild does require at least some setup. Were you also able to reproduce with `dpkg-buildpackage` or only with `sbuild`?

**Comment by bdrung** [comment]
As described above, I was only able to see all five failures inside the sbuild chroot. Only failure 1 was also caused when using dpkg-buildpackage directly. I tried to reproduce these issues outside of the schroot, but it always works there.

I can run the package build, enter the chroot after the test failure, and retrigger the test failures:
```
(stretch)user@host:/build/libgit2-2dpiwK/libgit2-0.25.1+git$ dh_auto_test --builddirectory=build-debian-release
[...]
  1) Failure:
repo::discover::discovering_repository_with_alternative_gitdir_succeeds [/build/libgit2-2dpiwK/libgit2-0.25.1+git/tests/repo/discover.c:36]
  Function call failed: (git_repository_discover(&found_path, start_path, 0, ceiling_dirs))
  error -3 - could not find repository from 'temprepo/discover.git/alternate_sub_repo/sub'

  2) Failure:
repo::discover::ceiling_should_not_affect_gitdir_redirection [/build/libgit2-2dpiwK/libgit2-0.25.1+git/tests/repo/discover.c:36]
  Function call failed: (git_repository_discover(&found_path, start_path, 0, ceiling_dirs))
  error -3 - could not find repository from 'temprepo/discover.git/alternate_sub_repo/sub'

  3) Failure:
repo::discover::discovery_starting_at_file_succeeds [/build/libgit2-2dpiwK/libgit2-0.25.1+git/tests/repo/discover.c:36]
  Function call failed: (git_repository_discover(&found_path, start_path, 0, ceiling_dirs))
  error -3 - could not find repository from 'temprepo/discover.git/sub_repo/file'

  4) Failure:
repo::open::force_bare [/build/libgit2-2dpiwK/libgit2-0.25.1+git/tests/repo/open.c:401]
  Function call failed: (git_repository_open_ext( &barerepo, "alternate/subdir/sub2", GIT_REPOSITORY_OPEN_BARE, ((void *)0)))
  error -3 - could not find repository from 'alternate/subdir/sub2'


    Start 2: libgit2_clar-cred_callback
2/4 Test #2: libgit2_clar-cred_callback ...............   Passed    0.00 sec
    Start 3: libgit2_clar-proxy_credentials_in_url
3/4 Test #3: libgit2_clar-proxy_credentials_in_url ....   Passed    0.00 sec
    Start 4: libgit2_clar-proxy_credentials_request
4/4 Test #4: libgit2_clar-proxy_credentials_request ...   Passed    0.00 sec

75% tests passed, 1 tests failed out of 4

Total Test time (real) =  10.32 sec

The following tests FAILED:
          1 - libgit2_clar (Failed)
```

Do you want me to test something or collect more information?

**Comment by pks-t** [comment]
I've not been able to reproduce the issues you have in a yakkety-amd64 schroot. But I've actually been able to reproduce different filesystem-related problems in that our tests with long filenames (e.g. `checkout::tree::extremely_long_file_name`) fail.

The issue seems to be related to how the schroot is actually set up, which is by default realized with a directory and overlays. The overlay is implemented via aufs, which restricts file names to 242 bytes in contrast to the usual 255 bytes common in other file systems. A workaround is to not set up the schroot with overlays but inside a file, instead, e.g. by using `mk-sbuild --type=file yakkety`.

You might try if a different type will also fix your issues, but I doubt that. My current reproduction steps:

```
mk-sbuild yakkety
git clone https://github.com/libgit2/libgit2
curl -LO http://http.debian.net/debian/pool/main/libg/libgit2/libgit2_0.24.5-1.debian.tar.xz
cd libgit2
git checkout v0.25.1
tar -xvf ../libgit2_0.24.5-1.debian.tar.xz
git archive HEAD | gzip >../libgit2_0.24.5.orig.tar.gz
sbuild -c yakkety-amd64
```

**Comment by bdrung** [comment]
My yakkety schroot configuration `/etc/schroot/chroot.d/yakkety` is configured this way:
```
[yakkety]
description=Ubuntu yakkety/amd64 autobuilder
groups=root,sbuild
root-groups=root,sbuild
profile=sbuild
type=directory
directory=/srv/chroot/yakkety
union-type=overlay
source-groups=root,sbuild
source-root-groups=root,sbuild
```
When I disable the usage of an overlay in the chroot (by commenting the `union-type` line), the test cases succeed. Note that I am using overlayfs instead of aufs which requires a more recent kernel than the deprecated aufs. Can you reproduce the test case failures when adding `union-type=overlay`?

**Comment by pks-t** [comment]
That's it, thanks! Just a wild guess in the dark what it'll be: we have checks inside our discovery code such that we do not search for repositories across file systems, which are (probably) realized by taking a look at `struct stat.st_dev`. I bet that the device changes when using overlayfs, where it does not for aufs.

Quick check -- yep, that's it. Try the following change:

```
diff --git a/tests/repo/discover.c b/tests/repo/discover.c
index abb7bd12b..eadd055e9 100644
--- a/tests/repo/discover.c
+++ b/tests/repo/discover.c
@@ -33,7 +33,7 @@ static void ensure_repository_discover(const char *start_path,
 	git_buf_attach(&resolved, p_realpath(expected_path, NULL), 0);
 	cl_assert(resolved.size > 0);
 	cl_git_pass(git_path_to_dir(&resolved));
-	cl_git_pass(git_repository_discover(&found_path, start_path, 0, ceiling_dirs));
+	cl_git_pass(git_repository_discover(&found_path, start_path, 1, ceiling_dirs));
 
 	cl_assert_equal_s(found_path.ptr, resolved.ptr);
 
diff --git a/tests/repo/open.c b/tests/repo/open.c
index 6114ad2e1..3239b6fec 100644
--- a/tests/repo/open.c
+++ b/tests/repo/open.c
@@ -398,7 +398,8 @@ void test_repo_open__force_bare(void)
 	cl_git_fail(git_repository_open_bare(&barerepo, "alternate/subdir/sub2"));
 
 	cl_git_pass(git_repository_open_ext(
-		&barerepo, "alternate/subdir/sub2", GIT_REPOSITORY_OPEN_BARE, NULL));
+		&barerepo, "alternate/subdir/sub2",
+		GIT_REPOSITORY_OPEN_BARE|GIT_REPOSITORY_OPEN_CROSS_FS, NULL));
 	cl_assert(git_repository_is_bare(barerepo));
 	git_repository_free(barerepo);
 }
 ```

aufs is broken due to file length limits, overlayfs is broken due to changing device numbers. 
So while I'm a bit hesitant to adjust our tests to search across filesystems, I guess it wouldn't do much harm anyway.

**Comment by pks-t** [comment]
Just took a quick stab at the overlayfs documentation:

```
This approach is 'hybrid' because the objects that appear in the
filesystem do not all appear to belong to that filesystem.  In many
cases an object accessed in the union will be indistinguishable
from accessing the corresponding object from the original filesystem.
This is most obvious from the 'st_dev' field returned by stat(2).
```

So it works as designed.

**Comment by bdrung** [comment]
Yes. The tests succeed when applying your patch.

**Comment by ethomson** [comment]
Closed via #4232
